lib/streamlit/runtime/pages_manager.py

```diff
@@ -144,10 +240,29 @@ def get_active_script_hash(self) -> PageHash:
     def set_active_script_hash(self, page_hash: PageHash):
         return self.pages_strategy.set_active_script_hash(page_hash)
 
-    def get_active_script(
+    def set_initial_script(
+        self, page_script_hash: PageHash, page_name: PageName
+    ) -> None:
+        return self.pages_strategy.set_initial_script(page_script_hash, page_name)
+
+    def get_initial_active_script(
         self, page_script_hash: PageHash, page_name: PageName
     ) -> Optional[PageInfo]:
-        return self.pages_strategy.get_active_script(page_script_hash, page_name)
+        return self.pages_strategy.get_initial_active_script(
+            page_script_hash, page_name
+        )
+
+    @contextlib.contextmanager
+    def run_with_active_hash(self, page_hash):
+        original_page_hash = self.get_active_script_hash()
+        self.set_active_script_hash(page_hash)
+        try:
+            yield
+        except Exception:
+            self.set_active_script_hash(original_page_hash)
+            raise
+        else:
+            self.set_active_script_hash(original_page_hash)
```

Comment:
Worth double checking, but I think writing

```python3
try:
    yield
finally:
    self.set_active_script_hash(original_page_hash)
```

should be equivalent (I think `finally` can sometimes essentially catch an exception, but only if a control flow statement is used within the finally block)